### PR TITLE
WIP: Fix M1 wheels, drop Python 3.6, lots of documentation

### PR DESCRIPTION
**Pull request by rileyjmurray**
## Description

I originally set out to fix issue #1604 for getting wheels to build on M1 macs. I investigated various possibilities (see my comment here: https://github.com/cvxpy/cvxpy/issues/1604#issuecomment-1046051882) and came to the conclusion that a simple use of environment markers in ``pyproject.toml`` should solve the problem. @akshayka @phschiele @h-vetinari please comment on if you think this will be sufficient!

I also resolved the deprecation warnings mentioned in #1654 and took steps to remove support for Python 3.6.

At some point, I got started on a documentation spree. @SteveDiamond @akshayka @bstellato: please review my changes to web documentation concerning the Reductions API. I basically took a stance that nothing in that realm is part of the public API. We need to take a stance there because we need to be able to determine what it means for a change to be backward-compatible. Definitely review the web documentation changes by building locally. I used a lot of ReStructuredText which made for clean web docs at some expense of messier source docs.

Thanks in advance for anyone's comments on this PR!

## Type of change
- [ ] New feature (backwards compatible)
- [ ] New feature (breaking API changes)
- [X] Bug fix
- [X] Other (Documentation, CI, ...)

## [Contribution checklist](https://www.cvxpy.org/contributing/index.html#contribution-checklist)
- [X] Add our license to new files.
- [X] Check that your code adheres to our coding style.
- [ ] Write unittests.
- [X] Run the unittests and check that they’re passing.
- [X] Run the benchmarks to make sure your change doesn’t introduce a regression.

### CODE REVIEW

File: doc/source/index.rst [hunk 1]
```diff
@@ -115,7 +116,7 @@ guide </contributing/index>` and join us `on Discord <https://discord.gg/4urRQeG
     :maxdepth: 3
     :hidden:
 
-    tutorial/index
+    User's Guide <tutorial/index>
```
Comment (akshayka) [inline doc/source/index.rst:119]: Thanks! I prefer this over tutorial. This is a very small nit: can you change "User's Guide" to "User Guide" everywhere? The latter is standard. (https://numpy.org/doc/stable/user/, https://en.wikipedia.org/wiki/User_guide)

File: cvxpy/atoms/elementwise/power.py [hunk 2]
```diff
@@ -57,16 +57,22 @@ class power(Elementwise):
...
+        the case. Specifically, ``power`` computes a rational approximation
         to ``p`` with a denominator up to ``max_denom``.
         Increasing ``max_denom`` can give better approximations.
-
         When ``p`` is an ``int`` or ``Fraction`` object, the approximation
-        is usually **exact**.
+        is usually *exact*. No such approximation
+        is used for DGP problems.
+
+        *UPDATE*: CVXPY supports exponential cone and power cone constraints.
```
Comment (akshayka) [inline cvxpy/atoms/elementwise/power.py:69]: Nice addition. Is the text "`*UPDATE*:`" necessary?

**Comment by akshayka** [comment]
Riley, this is fantastic. I've left very minor comments.

In the future we should take a pass over the "Advanced" section of the User Guide. It's a hodgepodge right now, some of which might be worth calling out in the non-advanced portion of the User Guide. In particular in might be worth highlighting the solve method options, as well as parametrized programs (especially when Philipp & Michael's massive improvements are merged in!)

File: cvxpy/atoms/elementwise/power.py (hunk 2 above)
Comment (rileyjmurray) [reply cvxpy/atoms/elementwise/power.py:69]: Thanks! Not really, no. I'll remove it.

File: doc/source/index.rst (hunk 1 above)
Comment (rileyjmurray) [reply doc/source/index.rst:119]: Done!

File: cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py [hunk 3]
```diff
@@ -71,7 +72,7 @@ def invert(self, solution, inverse_data):
         attr = {}
         if solution["solve_method"] == s.SCS:
             import scs
-            if StrictVersion(scs.__version__) < StrictVersion('3.0.0'):
+            if Version(scs.__version__) < Version('3.0.0'):
```
Comment (SteveDiamond) [inline cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py:75]: Can you comment on the change here? Is this worth adding a new dependency?

File: doc/source/updates/index.rst [hunk 4]
```diff
@@ -8,45 +8,50 @@ CVXPY's project maintainers currently provide support for CVXPY 1.2 and 1.1.
...
- important linear operators in quantum information.
-
+We review those features and the new features in CVXPY 1.2.0 below.
+
+Constraints and atoms
+~~~~~~~~~~~~~~~~~~~~~
+ * 1.2.0: added atoms for `partial trace <https://en.wikipedia.org/wiki/Partial_trace>`_ and partial transpose,
+   which are important linear operators in quantum information
+ * 1.2.0: updated ``kron`` so that either argument in ``kron(A, B)`` can be a non-constant affine Expression,
+   provided the other argument is constant. We previously required that ``A`` was constant.
```
Comment (SteveDiamond) [inline doc/source/updates/index.rst:28]: 1.2.0: added xexp atom

**Comment by SteveDiamond** [comment]
Fantastic stuff @rileyjmurray ! You should update the CI to drop python 3.6.

File: cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py (hunk 3 above)
Comment (rileyjmurray) [reply cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py:75]: Distutils is [deprecated](https://docs.python.org/3/whatsnew/3.10.html#distutils-deprecated) and recommends using ``packaging`` to access version comparisons. We already use setuptools though, and the Python foundation suggests using either setuptools or packaging in place of distutils. I'll check if we can get what we need with setuptools.

Comment (SteveDiamond) [reply cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py:75]: Ok, I didn't know distutils is deprecated.

File: doc/source/updates/index.rst (hunk 4 above)
Comment (rileyjmurray) [reply doc/source/updates/index.rst:28]: Good catch! Fixed.

**Comment by rileyjmurray** [comment]
Thanks for the reviews, @SteveDiamond and @akshayka! I've made most of the changes. I just need to investigate setuptools vs packaging per Steven's comment, and I need to fix the tests which are failing for weird error tolerance reasons.

@akshayka, I wholeheartedly agree that the "advanced" section of the tutorial (now User Guide) is due for some reorganization. I have some ideas on that front. Happy to work on them in a month or so.

File: cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py (hunk 3 above)
Comment (rileyjmurray) [reply cvxpy/reductions/solvers/conic_solvers/diffcp_conif.py:75]: @SteveDiamond I looked at what ``setuptools`` offers for checking versions and it's not a practical replacement for ``packaging.version``. I just made a very simple version class that works for how we used ``distutils.StrictVersion``. This way we don't need to add ``packaging`` as a dependency.

**Comment by rileyjmurray** [comment]
This also "fixes" #1672 by adding more informative error messages.

**Review by SteveDiamond — APPROVED** [review]
LGTM

**Comment by rileyjmurray** [comment]
@akshayka @SteveDiamond @bodono some tests were failing with SCS 3.2. I went back and found that our SCS interface was only attempting re-solves with Anderson acceleration disabled for SCS 2.x. In order to get tests to pass, I had to enable re-solves with all versions of SCS >= 2.0. I added a warning message which is raised when the re-solve occurs, which largely resolves #1164. Note that the warning is raised regardless of whether or not ``verbose=True``. How do each of you feel about this approach?

 I also had to make termination criteria more stringent in some tests that used SCS.

**Comment by rileyjmurray** [comment]
Okay turns out SCS 3.0 disables Anderson acceleration by default. I think I got confused with the source of errors when I initially ran tests with SCS 2.1.4 locally, say SCS 3.0 errors on GH Actions, and then updated my local install to 3.0. Give me some time to figure this out.

**Comment by h-vetinari** [comment]
> @h-vetinari please comment on if you think this will be sufficient!

Thanks for the ping. I'll happily try this in a new release, but this isn't a big issue for conda-forge. Carrying a patch to remove (currently) `pyproject.toml` isn't amazing, but also not terrible either. If it's not needed anymore, all the better, but won't make-or-break the workflow for us.

**Comment by rileyjmurray** [comment]
@bodono 

> SCS 3.x hasn't disabled AA by default, but it should be much, much more stable and cheaper to apply, so I'm surprised to hear you are having issues with it

Okay, I misunderstood the SCS web documentation. It seemed like the default value stated there was ``acceleration_lookback=0``. But I looked at some solver output when cvxpy didn't specify AA lookback and it said lookback=10. 

All in all, it seems I misattributed the various failures to AA. There were two things that went wrong in SCS testing:

1. SCS 3.2 fails some accuracy tests that previous versions of SCS 3.x passed. Minor changes to accuracy parameter ``eps`` in calls to ``prob.solve()`` fixed this.
2. Our test ``TestSCS.test_scs_options`` provided an usual parameter of ``alpha=1e-4``, which caused SCS to diverge. This had nothing to do with AA. Similar divergence was occurring in SCS 3.1 and 3.0, but the iteration limit was so low that SCS wasn't able to diverge "too much" and happened to pass the tests which had very generous tolerances. The solution here was just to fix ``alpha`` to a more reasonable value like 1.0 or 1.4.

Since neither of these was due to AA we have a decision to make: when encountering a solver failure with SCS 3.x and default AA, should cvxpy automatically re-solve the problem with AA turned off? That's the approach we take with 2.x. @bodono since you say AA is much more reliable I'm open to putting this behind us, and only doing re-solves with SCS 2.x is installed.

@SteveDiamond, @phschiele @akshayka your opinions on this would also be appreciated.

**Comment by phschiele** [comment]
@rileyjmurray In my view, since the current behaviour did not lead to issues being reported that were caused by AA, as well as the additional information that it is generally much more stable under 3.x, I think it's ok to only resolve under 2.x.

**Comment by SteveDiamond** [comment]
I agree with Philipp, if we can avoid resolving that's preferable.

**Comment by rileyjmurray** [comment]
That's three in favor! I expect Akshay and Brendan will either agree or have only mild dissent. I'll make the necessary final changes and then merge. Akshay and Brendan: we can of course revisit if you feel strongly about this.

**Comment by bodono** [comment]
> Okay, I misunderstood the SCS web documentation. It seemed like the default value stated there was acceleration_lookback=0. But I looked at some solver output when cvxpy didn't specify AA lookback and it said lookback=10.

Sorry, the documentation was wrong. I fixed that here: https://github.com/cvxgrp/scs/pull/212

**Comment by rileyjmurray** [comment]
Well, shoot. I merged it in without changing the "WIP" title. Ah well. We're done here!

**Pull request merged**